modules/ml/src/svmsgd.cpp

```diff
@@ -0,0 +1,510 @@
+    // Stochastic gradient descent SVM
+    for (int iter = 0; (iter < maxCount) && (err > epsilon); iter++)
+    {
+        int randomNumber = rng.uniform(0, extendedTrainSamplesCount);             //generate sample number
+
+        Mat currentSample = extendedTrainSamples.row(randomNumber);
+
+        float stepSize = params.initialStepSize * std::pow((1 + params.marginRegularization * params.initialStepSize * (float)iter), (-params.stepDecreasingPower));    //update stepSize
+
+        updateWeights( currentSample, isPositive(trainResponses.at<float>(randomNumber)), stepSize, extendedWeights );
+
+        //average weights (only for ASGD model)
+        if (params.svmsgdType == ASGD)
+        {
+            averageExtendedWeights = ((float)iter/ (1 + (float)iter)) * averageExtendedWeights  + extendedWeights / (1 + (float) iter);
+            err = norm(averageExtendedWeights - previousWeights);
+            averageExtendedWeights.copyTo(previousWeights);
+        }
+        else
+        {
+            err = norm(extendedWeights - previousWeights);
+            extendedWeights.copyTo(previousWeights);
+        }
+    }
+
+    if (params.svmsgdType == ASGD)
+    {
+        extendedWeights = averageExtendedWeights;
+    }
+
+    Rect roi(0, 0, featureCount, 1);
+    weights_ = extendedWeights(roi);
+    weights_ *= multiplier;
+
+    CV_Assert((params.marginType == SOFT_MARGIN || params.marginType == HARD_MARGIN) && (extendedWeights.type() ==  CV_32FC1));
+
+    if (params.marginType == SOFT_MARGIN)
+    {
+        shift_ = extendedWeights.at<float>(featureCount) - static_cast<float>(weights_.dot(average));
+    }
+    else
+    {
+        shift_ = calcShift(trainSamples, trainResponses);
+    }
+
+    return true;
+}
+
+float SVMSGDImpl::predict( InputArray _samples, OutputArray _results, int ) const
+{
+    float result = 0;
+    cv::Mat samples = _samples.getMat();
+    int nSamples = samples.rows;
+    cv::Mat results;
+
+    CV_Assert( samples.cols == weights_.cols && samples.type() == CV_32FC1);
+
+    if( _results.needed() )
+    {
+        _results.create( nSamples, 1, samples.type() );
+        results = _results.getMat();
+    }
+    else
+    {
+        CV_Assert( nSamples == 1 );
+        results = Mat(1, 1, CV_32FC1, &result);
+    }
+
+    for (int sampleIndex = 0; sampleIndex < nSamples; sampleIndex++)
+    {
+        Mat currentSample = samples.row(sampleIndex);
+        float criterion = static_cast<float>(currentSample.dot(weights_)) + shift_;
+        results.at<float>(sampleIndex) = (criterion >= 0) ? 1.f : -1.f;
+    }
+
+    return result;
+}
+
+bool SVMSGDImpl::isClassifier() const
+{
+    return (params.svmsgdType == SGD || params.svmsgdType == ASGD)
+            &&
+            (params.marginType == SOFT_MARGIN || params.marginType == HARD_MARGIN)
+            &&
+            (params.marginRegularization > 0) && (params.initialStepSize > 0) && (params.stepDecreasingPower >= 0);
```

Comment:
When params.stepDecreasingPower == 0, step size is a constant. It is the valid value.